posts/2021/10/numpy_benchmarking_blog.md

```diff
@@ -0,0 +1,165 @@
+</p>
+
+<!-- TEASER_END -->
+
+## My Experience
+My work was broadly divided majorly into the following categories:
+
+- **N-Body Problem**: The [N-Body problem](https://en.wikipedia.org/wiki/N-body_problem) is one of the most famous 
+and universally accepted problems for benchmarking. 
+I was given this as a problem statement to work on the project. 
+I started my work with a theoretical understanding of the problem. 
+I took reference from [Pierre Augier] remarkable work on 
+[n-body](https://github.com/paugier/nbabel) problem. I'd love to thank him.
+It was a fun learning part for me to connect the scientific part with the programming world. 
+I implemented the N-Body problem in Python, C, and C++.
+
+- **Compiled Methods**: This part was the most exciting part of the project. 
+I got introduced to various compiled methods like [Cython](https://cython.readthedocs.io/en/latest/), 
+[Pythran](https://pythran.readthedocs.io/en/latest/), [Numba](http://numba.pydata.org/), [Transonic](https://transonic.readthedocs.io/en/latest/). 
+It was the first time when I got to know about accelerators. 
+I visited their documentations, GitHub, and searched popular blogs about compiled methods. 
+I loved playing with them. I read the theory part, looked into the examples, 
+and implemented them in my editor. It was a great learning experience for me 
+to get familiar with compiled methods. 
+I implemented the jitted compilation mode in Pythran and Numba for benchmarking via Transonic's support.
+
+- **Visualization**: I used the [Matplotlib](https://matplotlib.org/) library for visualizing the benchmarking results. 
+I tried various plots to verify which one suits best, 
+like scatter plots, box plots, line charts, the combination of the scatter plots and line charts, etc. 
+But these were not good to go. 
+Those plots either lacked clarity or were not capable of providing significant results. 
+We finalized decided on horizontally stacked bar charts.
+
+- **Model Optimization**: Model Optimization was one of the most exciting parts for me to work. 
+I like playing with codes. The main task was to ensure 
+we were obtaining similar results in all the implemented algorithms. 
+I revisited all the code I had executed earlier. At this stage, 
+I was able to find out errors in my code and had an idea to improve it. 
+The final aim was to achieve the same results in each step at a minimal time. 
+Steps I followed to attain it:
+	
+	- Initially, I played around with the library functions to check out which library function gave the best results.
+	
+	- I then turned my focus to reduce the number of loops. 
+And I'll say hats off to the *Vectorized Approach* of the NumPy. 
+NumPy achieved more than 10% than Python and 
+equivalent speed to compiled methods like C++ and Numba.
+	
+	- The only task left was to verify whether we got the same results in all the cases. 
+Initially, I wanted to make my code as small as possible. 
```

Comment:
But was it really faster? The difference in speed between `np.add(a, b)` and `a + b` was negligible, if measurable at all.
```suggestion
Initially, I wanted to make my code as fast as possible. 
```

Reply:
> But was it really faster? The difference in speed between np.add(a, b) and a + b was negligible, if measurable at all.

Yes, you are correct those differences were negligible and were varying. 

Here I was talking about the non-trivial approach I used earlier.
This one: https://github.com/numpy/numpy.org/pull/461#discussion_r705694228

Does that make sense?